### PR TITLE
feat(plugin): PictureInPicture

### PR DESCRIPTION
**Pull request by lumap**
This plugin adds a picture in picture button for videos next to the Download button.
TODO:
- ~~Make the SVG fit more with the others color-wise~~
- ~~Find a way to remove the ts-ignore in the onClick event~~
- ~~Make the hover tooltip work~~
- ~~Test more?~~
I think most of the essential stuff is done

### CODE REVIEW

File: src/plugins/pictureInPicture.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import { React, Tooltip } from "@webpack/common";
+
+export default definePlugin({
+    name: "PitureInPicture",
+    description: "Adds picture in picture to videos (next to the Download button)",
+    authors: [Devs.Lumap],
+    patches: [{
+        find: ".onRemoveAttachment,",
+        replacement: {
+            match: /\.nonMediaAttachment.{0,10}children:\[(.),/gm,
```
Comment (Vendicated) [inline src/plugins/pictureInPicture.tsx:18]: unless you deliberately mean to replace multiple places
```suggestion
            match: /\.nonMediaAttachment.{0,10}children:\[(.),/,
```

Comment (Vendicated) [reply src/plugins/pictureInPicture.tsx:18]: use \i, not .

File: src/plugins/pictureInPicture.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                    }}
+                    onClick={e => {
+                        try {
+                            // @ts-ignore
+                            e.target.parentNode.parentNode.querySelector("video").requestPictureInPicture();
+                        } catch { }
+                    }}
+                >
+                    <svg width="24px" height="24px" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg" style={{ pointerEvents: "none" }}>
+                        <g>
```
Comment (Vendicated) [inline src/plugins/pictureInPicture.tsx:47]: remove this useless group

File: src/plugins/pictureInPicture.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                        paddingRight: "4px",
+                    }}
+                    onClick={e => {
+                        try {
+                            // @ts-ignore
+                            e.target.parentNode.parentNode.querySelector("video").requestPictureInPicture();
+                        } catch { }
+                    }}
+                >
+                    <svg width="24px" height="24px" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg" style={{ pointerEvents: "none" }}>
```
Comment (Vendicated) [inline src/plugins/pictureInPicture.tsx]:
```suggestion
                    <svg width="24px" height="24px" viewBox="0 0 24 24">
```

File: src/plugins/pictureInPicture.tsx [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    renderPiPButton() {
+        return <Tooltip text={"Enable Picture in Picture"}>
+            {/* {this.PiPButtonContent} */}
+            {({ onMouseLeave, onMouseEnter }) => (
+                <div
+                    onMouseLeave={onMouseLeave}
+                    onMouseEnter={onMouseEnter}
+                    role="button"
+                    aria-label={"Enable Picture in Picture"}
+                    aria-disabled={false}
```
Comment (Vendicated) [inline src/plugins/pictureInPicture.tsx:31]:
```suggestion
```

File: src/plugins/pictureInPicture.tsx [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    authors: [Devs.Lumap],
+    patches: [{
+        find: ".onRemoveAttachment,",
+        replacement: {
+            match: /\.nonMediaAttachment.{0,10}children:\[(.),/gm,
+            replace: "$&$1&&$self.renderPiPButton(),"
+        }
+    }],
+    renderPiPButton() {
+        return <Tooltip text={"Enable Picture in Picture"}>
```
Comment (Vendicated) [inline src/plugins/pictureInPicture.tsx:23]:
```suggestion
        return <Tooltip text="Enable Picture in Picture">
```

File: src/plugins/pictureInPicture.tsx [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            {/* {this.PiPButtonContent} */}
+            {({ onMouseLeave, onMouseEnter }) => (
+                <div
+                    onMouseLeave={onMouseLeave}
+                    onMouseEnter={onMouseEnter}
+                    role="button"
+                    aria-label={"Enable Picture in Picture"}
```
Comment (Vendicated) [inline src/plugins/pictureInPicture.tsx:30]: iirc, the tooltip props contain an aria Label, so if you just spread the tooltip props info your div it should just work

**Review by Vendicated — CHANGES_REQUESTED** [review]

**Comment by lumap** [comment]
issue i noticed: switching channels when a video is playing in PiP doesn’t close the PiP window, it just pauses the video until you close it

**Comment by lumap** [comment]
next up: ff the video in PiP to the same time at the og video if it’s playing and pause the og video afterwards

File: src/plugins/pictureInPicture.tsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import { Settings } from "@api/Settings";
+import { Devs } from "@utils/constants";
+import definePlugin, { OptionType } from "@utils/types";
+import { React, Tooltip } from "@webpack/common";
+
+export default definePlugin({
+    name: "PictureInPicture",
+    description: "Adds picture in picture to videos (next to the Download button)",
+    authors: [Devs.Lumap],
+    options: {
```
Comment (Vendicated) [inline src/plugins/pictureInPicture.tsx:16]: options are deprecated. use settings coupled with definePluginSettings

File: src/plugins/pictureInPicture.tsx [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+            type: OptionType.BOOLEAN,
+            default: true,
+            restartNeeded: false
+        }
+    },
+    patches: [
+        {
+            find: ".onRemoveAttachment,",
+            replacement: {
+                match: /\.nonMediaAttachment.{0,10}children:\[(\i),/gm,
```
Comment (Vendicated) [inline src/plugins/pictureInPicture.tsx:28]:
```suggestion
                match: /\.nonMediaAttachment.{0,10}children:\[(\i),/,
```
don't add flags unless you need them

File: src/plugins/pictureInPicture.tsx [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+                replace: "$&$1&&$self.renderPiPButton(),"
+            },
+        },
+    ],
+    renderPiPButton() {
```
Comment (Vendicated) [inline src/plugins/pictureInPicture.tsx:33]: wrap in ErrorBoundary for error safety

**Comment by lumap** [comment]
> next up: ff the video in PiP to the same time at the og video if it’s playing and pause the og video afterwards

implemented this as well as all of ven's review suggestions

File: src/plugins/pictureInPicture.tsx [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+                        {...tooltipProps}
+                        role="button"
+                        style={{
+                            cursor: "pointer",
+                            paddingTop: "4px",
+                            paddingLeft: "4px",
+                            paddingRight: "4px",
+                        }}
+                        onClick={e => {
+                            const oldVid = e.currentTarget.parentNode!.parentNode!.querySelector("video")!;
```
Comment (Vendicated) [inline src/plugins/pictureInPicture.tsx:50]: please add some comments explaing why stuff is done. for example, why do you conditionally display: none it

**Comment by Vendicated** [comment]
thank you!

**Review by Vendicated — APPROVED** [review]

**Comment by Cyanide0210** [comment]
Can we also add this to streams please?

**Comment by castdrian** [comment]
video embeds within richembeds could also use at least the looped playback

**Pull request merged**